### DISCUSSION TITLE
[JS 101]  變數宣告問題

### DISCUSSION THREAD

**Discussion by s103071049** [Lidemy/forum#62] [其他問題]
[JS 101] 寫練習七字串反轉時，發現變數宣告的位置不同，會影響印出的結果。查完資料後還是有些疑惑，Local variable doesn't provide data sharing whereas Global variable provides data sharing，我可以理解成局部變數只會在函式內起作用，所以我重新呼叫這個函式，對局部變數而言它的值就會重新運算。對全域變數而言，因為data sharing所以上次函式呼叫完的值會被保存，下次的值相當於是append進去的?這樣對於local global variable的理解對嗎?

變數放在function內 (local variable)
``` js
function reverse(str){
	var result=""
	for(var i=str.length-1;i>=0;i--){
		result+=str[i]
		
	}return result
}

console.log(reverse("12345aa")) //aa54321

console.log(reverse("abcd")) //dcda

```

變數放在function外  (globall variable)
``` js

var result=""
function reverse(str){
	
	for(var i=str.length-1;i>=0;i--){
		result+=str[i]
		
	}return result
}

console.log(reverse("12345aa")) //aa54321

console.log(reverse("abcd")) //aa54321dcda

```

**Comment by aszx87410** [comment]
> 我可以理解成局部變數只會在函式內起作用，所以我重新呼叫這個函式，對局部變數而言它的值就會重新運算。

這個沒錯

> 對全域變數而言，因為data sharing所以上次函式呼叫完的值會被保存，下次的值相當於是append進去的?

前半段是對的，後半段有點怪怪的，append 進去是因為你寫 `result+=`

----------

用 var 宣告的變數他的生命範圍就是宣告的 function 之內，生命範圍的意思是一旦脫離了 function，就沒有辦法存取到。你每次進入 function 的時候可以想成都會重新宣告一個變數，而不是用之前的

舉例來說你第一個範例中呼叫了兩次 reverse()，第一次裡面的 result 跟第二次裡面的 result 會是不同的變數，存在不同的記憶體位置

第二個範例把它宣告在 function 外面的話，以這個例子而言就會像是全域變數，所以你呼叫了兩次 reverse，裡面的 result 都會是同一個變數

但其實很多時候宣告在哪裡不是重點，重點是第二個範例你可以在 reverse function 的第一行加上 `result = ''` 把 result 清空，這樣也不會有 bug

**Reply by s103071049** [reply]
理解了!!!，老師謝謝你ˊˇˋ
